networking_aci/plugins/ml2/drivers/mech_aci/common.py

```diff
@@ -135,11 +137,95 @@ def get_hosts_on_network(self, context, network_id, level=None, with_segment=Fal
                 hosts.add(host)
         return hosts
 
+    def get_hosts_on_physnet(self, context, physical_network, level=None, with_segment=False, with_segmentation=False):
+        """Get all binding hosts (from host or binding_profile) present on a network"""
+        fields = [ml2_models.PortBinding.host, ml2_models.PortBinding.profile]
+        if with_segment:
+            fields.append(ml2_models.PortBindingLevel.segment_id)
+        if with_segment:
+            fields.append(segment_models.NetworkSegment.segmentation_id)
+        query = context.session.query(*fields)
+        query = query.join(ml2_models.PortBindingLevel,
+                           ml2_models.PortBinding.port_id == ml2_models.PortBindingLevel.port_id)
+        query = query.join(segment_models.NetworkSegment,
+                           ml2_models.PortBindingLevel.segment_id == segment_models.NetworkSegment.id)
+        query = query.filter(segment_models.NetworkSegment.physical_network == physical_network)
+        if level is not None:
+            query = query.filter(ml2_models.PortBindingLevel.level == level)
+
+        hosts = set()
+        for entry in query.all():
+            host = get_host_from_profile(entry.profile, entry.host)
+            if with_segment or with_segmentation:
+                row = [host]
+                if with_segment:
+                    row.append(entry.segment_id)
+                if with_segmentation:
+                    row.append(entry.segmentation_id)
+                hosts.add(tuple(row))
+            else:
+                hosts.add(host)
+        return hosts
+
     def get_segment_ids_by_physnet(self, context, physical_network):
         query = context.session.query(segment_models.NetworkSegment.id)
         query = query.filter(segment_models.NetworkSegment.physical_network == physical_network)
         return [seg.id for seg in query.all()]
 
+    def get_ports_on_network_by_physnet_prefix(self, context, network_id, physical_network_prefix):
+        # get all ports for a network that are on a segment with a physnet prefix
+        fields = [
+            segment_models.NetworkSegment.id, segment_models.NetworkSegment.physical_network,
+            models_v2.Port.id, models_v2.Port.project_id
+        ]
+        query = context.session.query(*fields)
+        query = query.filter(models_v2.Port.network_id == network_id)
+        query = query.join(ml2_models.PortBindingLevel, ml2_models.PortBindingLevel.port_id == models_v2.Port.id)
+        query = query.join(segment_models.NetworkSegment,
+                           ml2_models.PortBindingLevel.segment_id == segment_models.NetworkSegment.id)
+        query = query.filter(segment_models.NetworkSegment.physical_network.like('{}%'.format(physical_network_prefix)))
+
+        result = []
+        for entry in query.all():
+            result.append({
+                'segment_id': entry[0],
```

Comment:
Why is this `entry` a tuple or list while the previous `entry` in `get_hosts_on_physnet()` provides attribute access? Or do both and we just handle things differently here?

Reply:
SQLAlchemy provides objects that allow us to do both. Via index I can access the fields in the order I have requested them and with attribute access I can access them by name. If they overlap though I don't know which member I'm getting (it's consistent, but I don't want to do any guesswork), therefore I go by index here.